paddle/framework/attribute_reader.h

```diff
@@ -0,0 +1,193 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include <google/protobuf/map.h>
+#include <paddle/framework/attribute.pb.h>
+#include <paddle/framework/enforce.h>
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
+namespace paddle {
+namespace framework {
+using AttributeMap = google::protobuf::Map<std::string, Attribute>;
+class AttributeReader {
+ public:
+  explicit AttributeReader(const AttributeMap& attrs) : attrs_(attrs) {}
+
+  /**
+   * @brief Contains a attribute with name and type T.
+   *
+   * The example code like
+   * @code{cpp}
+   * AttributeReader reader;
+   *
+   * assert(reader.Contain<int>("SomeIntValue")==true);
+   * assert(reader.Contain<float>("SomeIntValue")==false);
+   * assert(reader.Contain<std::vector<int>>("SomeIntList")==true);
+   * @endcode{cpp}
+   *
+   * @tparam T Attribute Type, could be {int, float, string and std::vector of
+   * them}.
+   * @param name attribute name
+   * @return true if contain an attribute with name and type T, false if Type
+   * mismatch or not contains that name.
+   */
+  template <typename T>
+  bool Contain(const std::string& name) const;
+
+  /**
+   * @brief Get Attribute value. Not support std::vector. If want to return a
+   * std::vector, use `GetArray`
+   * @tparam T could be int, float, string.
+   * @param name attribute name.
+   * @return Value
+   * @throw If attribute is not found or type mismatch, an EnforceNotMet will
+   * throw
+   */
+  template <typename T>
+  T Get(const std::string& name) const;
+
+  /**
+   * @brief Get Attribute array values.
+   * @tparam T could be int, float, string.
+   * @param name attribute name.
+   * @param vec the return vector. Must be empty.
+   * @throw If attribute is not found, or type mismatch, or vec is not empty, an
+   * EnforceNotMet will throw
+   */
+  template <typename T>
+  void GetArray(const std::string& name, std::vector<T>* vec) const;
+
+ private:
+  const AttributeMap& attrs_;
+};
+
+/// Implementation of Contain
+namespace details {
```

Comment:
Because we need define some function that should not be used by the user but can only be defined in the header file.

So mark them in `namespace details` could tell the user not use them directly.